Installation issue: gcc@8.,3.0 with centos7 default compiler

### Steps to reproduce the issue

```console
$ spack install gcc@8.3.0
...
  >> 2426    checking for off_t... checking size of long long... configure: err
             or: in `spack-stage-gcc-8.3.0-6ypg
             gmswzvrzgw3lq6tovkv23mect6ee/spack-src/spack-build/gcc':
  >> 2427    configure: error: cannot compute sizeof (long long)
```

### Information on your system

$ spack debug report
* **Spack:** 0.15.0
* **Python:** 3.6.8
* **Platform:** linux-centos7-broadwell

### Additional information

Related to #3943 and #1866 which were not properly fixed but *avoided*.

PR #9024 solves this issue. (and I think +binutils should become the default).

Maintainer is @michaelkuhn .

### General information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [ ] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

I'm having the same problem, using `spack@develop` (b3bc538df649c5b6bfaba697b70cf874da454c26 from `2020-07-06`) on all the following platforms:
* CentOS 7, CentOS 8 (x86_64 and ppc64le)
* RHEL 7, RHEL 8 (x86_64 and ppc64le)

I'm trying to build gcc@7.3.0 using the system compiler in each of these environments. It is failing in all cases with same error. 

<b>CentOS 7 using system compiler gcc@4.8.5</b>
```
$> spack install gcc@7.3.0+strip target=x86_64
...
==> Error: Failed to install gcc due to ChildError: ProcessError: Command exited with status 2:
    'make' '-j16'
5 errors found in build log:
     2416    checking for size_t... 0
     2417    checking size of long... yes
     2418    checking for ssize_t... 0
     2419    checking for long long... yes
     2420    checking for uintptr_t... yes
     2421    checking size of long long... yes
  >> 2422    checking for ptrdiff_t... configure: error: in `/tmp/root/spack-stage/spack-stage-gcc-7.3.0-jkibfh6hlr3lyzsmcqbkxvqvhmnxgk32/spack-src/spack-build/gcc':
  >> 2423    configure: error: cannot compute sizeof (long long)
     2424    See `config.log' for more details.
  >> 2425    make[2]: *** [configure-stage1-gcc] Error 77
     2426    make[2]: *** Waiting for unfinished jobs....
     2427    yes
     2428    checking for uint64_t... yes
     2429    checking whether struct tm is in sys/time.h or time.h... time.h
     2430    checking size of int... 4
     2431    checking size of long... 8

     ...

     2481    configure: creating ./config.status
     2482    config.status: creating Makefile
     2483    config.status: creating config.h
     2484    config.status: executing depdir commands
     2485    mkdir -p -- .deps
     2486    make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-gcc-7.3.0-jkibfh6hlr3lyzsmcqbkxvqvhmnxgk32/spack-src/spack-build'
  >> 2487    make[1]: *** [stage1-bubble] Error 2
     2488    make[1]: Leaving directory `/tmp/root/spack-stage/spack-stage-gcc-7.3.0-jkibfh6hlr3lyzsmcqbkxvqvhmnxgk32/spack-src/spack-build'
  >> 2489    make: *** [all] Error 2
```

Here is full build log: [spack-log.txt](https://github.com/spack/spack/files/4879884/spack-log.txt)

Same type of error w the other OS: CentOS 8, RHEL 7, RHEL 8.

@becker33

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -115,7 +115,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     depends_on('zstd', when='@10:')
     depends_on('iconv', when='platform=darwin')
     depends_on('gnat', when='languages=ada')
-    depends_on('binutils~libiberty', when='+binutils')
+    depends_on('binutils~libiberty', when='+binutils', type=('build', 'run'))
     depends_on('zip', type='build', when='languages=java')
     depends_on('cuda', when='+nvptx')
 

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -373,8 +373,6 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
         # enable appropriate bootstrapping flags
         stage1_ldflags = str(self.rpath_args)
         boot_ldflags = stage1_ldflags + ' -static-libstdc++ -static-libgcc'
-        if '%gcc' in spec:
-            stage1_ldflags = boot_ldflags
         options.append('--with-stage1-ldflags=' + stage1_ldflags)
         options.append('--with-boot-ldflags=' + boot_ldflags)
 

--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -115,7 +115,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     depends_on('zstd', when='@10:')
     depends_on('iconv', when='platform=darwin')
     depends_on('gnat', when='languages=ada')
-    depends_on('binutils~libiberty', when='+binutils', type=('build', 'run'))
+    depends_on('binutils~libiberty', when='+binutils', type=('build', 'link', 'run'))
     depends_on('zip', type='build', when='languages=java')
     depends_on('cuda', when='+nvptx')
 

